src/lib.rs

```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
     }
 }
 
-fn concat_slice<'a>(a: &'a [u8], b: &'a [u8]) -> Result<&'a [u8], Error> {
+/// Concatenate two slices if they are adjacent.
+///
+/// If two slices are adjacent to each other in memory, this function
+/// concatenates both, creating a single longer slice. Note that slices of
+/// zero-sized types (ZST) are never considered adjacent. Otherwise it would be
+/// possible to concatenate a slice to itself.
+///
+/// # Errors
+///
+/// Returns `Err` if the two slices aren't adjacent, `a` is after `b`, or if the
+/// result is too long to be represented as a slice (size in bytes is larger
+/// than `isize::MAX`).
+///
+/// When T is a ZST then returns `Err(TooLong)` if the total length would overflow
+/// `usize` and `Err(NotAdjacent)` otherwise.
+///
+/// # Examples
+///
+/// Correct usage:
+///
+/// ```rust
+/// # use str_concat::concat_slice;
+/// let s = b"0123456789";
+/// assert_eq!(b"0123456", concat_slice(&s[..5], &s[5..7]).unwrap());
+/// ```
+///
+/// Non-adjacent byte slices:
+///
+/// ```rust
+/// # use str_concat::{concat_slice, Error};
+/// let s = b"0123456789";
+/// assert_eq!(Err(Error::NotAdjacent), concat_slice(&s[..5], &s[6..7]))
+/// ```
+///
+pub fn concat_slice<'a, T>(a: &'a [T], b: &'a [T]) -> Result<&'a [T], Error> {
     let a_ptr = a.as_ptr();
     let b_ptr = b.as_ptr();
 
     let a_len = a.len();
     let b_len = b.len();
 
+    if mem::size_of::<T>() == 0 {
+        // Not safety critical but we check that the length is as expected.
+        if a_len.checked_add(b_len).is_none() {
```

Comment:
`isize::max_value()` is the bound of the byte length. The byte length of a ZST slice is always `0`. The safety requirement is thus trivially fulfilled and the check here only checks if the result would be *correct*.

Also, it is possible to fit two ZST slices with length `usize::max_value()` into memory but we can surely not concatenate them. If we were to relax the requirements and allow concatenation of two ZST slices, ever, it might be beneficial to already return that error variant and have the check.

Reply:
But the checks are in a different order than the non-zero checks so I see your point.